inc/macro_utils/macro_utils.h

```diff
@@ -97,82 +97,136 @@ MU_IF(X, "true", "false") => "true"
+    const char* MU_C3(MU_, enumName, _ToString)(enumName value);
+
 #define MU_DEFINE_ENUMERATION_CONSTANT_AS_WIDESTRING(x) MU_C2(L, MU_TOSTRING(x)) , 
 #define MU_DEFINE_ENUMERATION_CONSTANT_AS_STRING(x) MU_TOSTRING(x) , 
+
+typedef enum MU_ENUM_VALUE_CONTAINS_EQUAL_TAG
+{
+    MU_ENUM_VALUE_CONTAINS_EQUAL_UNKNOWN,
+    MU_ENUM_VALUE_CONTAINS_EQUAL_YES,
+    MU_ENUM_VALUE_CONTAINS_EQUAL_NO
+} MU_ENUM_VALUE_CONTAINS_EQUAL;
+
+// variable definition indicating if an enum value label contains or not equal
+#define MU_INTERNAL_DEFINE_ENUM_HAS_EQ_VAR(argCount, enumValue) \
+    static MU_ENUM_VALUE_CONTAINS_EQUAL MU_C2(enum_value_has_equal_, argCount) = MU_ENUM_VALUE_CONTAINS_EQUAL_UNKNOWN; \
+
+// determine whether an enum value label has an equal or not
+#define MU_INTERNAL_COMPUTE_ENUM_HAS_EQ(argCount, enumValue) \
+    switch (MU_C2(enum_value_has_equal_, argCount)) \
+    { \
+        default: \
+        case MU_ENUM_VALUE_CONTAINS_EQUAL_UNKNOWN: \
+            if (strchr(MU_TOSTRING(MU_C2(my_, enumValue)), '=') != NULL) \
+            { \
+                MU_C2(enum_value_has_equal_, argCount) = MU_ENUM_VALUE_CONTAINS_EQUAL_YES; \
+                has_equal = 1; \
+            } \
+            else \
+            { \
+                MU_C2(enum_value_has_equal_, argCount) = MU_ENUM_VALUE_CONTAINS_EQUAL_NO; \
+                has_equal = 0; \
+            } \
+            break; \
+        case MU_ENUM_VALUE_CONTAINS_EQUAL_YES: \
+            has_equal = 1; \
+            break; \
+        case MU_ENUM_VALUE_CONTAINS_EQUAL_NO: \
+            has_equal = 0; \
+            break; \
+    } \
+
+// This is the bulk of the _ToString for an enum
+// The algorithm is:
+// - Go through each enum value
+//   - Determine if it contains an equal by searching '=' in the enum value string
+//     This value is cached. Threading is not taken into account, if a dirty read happens and an unknown value is obtained
+//     the code will simply take the worse case and take a performance hit in recomputing whether an equal is in the string
+//   - Determine the actual value of the enum value that we need to compare against:
+//     - Have a fake variable that is either initialized because there is an equal in the enum label or it is initialized to 0 
+//       because the fake variables are static
+//     - If the enum value contains an equal then take as current enum value whatever is in the fake variable
+//     - If the enum value does not contain an equal then compute the current enum value as the previous enum value plus 1
+//   - If the currently computed enum value matches the value requested to be stringified by the user then return the string to the user
+#define MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID_COMMON_CODE(enumName, ...) \
+    { \
+        MU_FOR_EACH_1_COUNTED(MU_INTERNAL_DEFINE_ENUM_HAS_EQ_VAR, __VA_ARGS__) \
+        MU_FOR_EACH_1_KEEP_1(MU_INTERNAL_DEFINE_ENUM_VAR, enumName, __VA_ARGS__) \
+        int has_equal = 0; \
```

Comment:
>has_equal [](http://example.com/codeflow?start=12&length=9)

with the removal of MU_INTERNAL_COMPUTE_ENUM_HAS_EQ who reads this value?

Reply:
MU_INTERNAL_COMPUTE_ENUM_HAS_EQ cannot be removed. That part was fixed (tests were added and now it is used.

---
In reply to: [632153203](https://github.com/Azure/macro-utils-c/pull/30#discussion_r632153203) [](ancestors = 632153203)